pbdutil.1
.\" $Id: pbdutil.1,v 1.7 2018/05/01 08:39:02 rok Exp $
.\"
.Dd -
.Dt PBDUTIL 1
.Os "Mac OS X"
.Sh NAME
.Nm pbdutil
.Nd read and write the Pasteboard
.Sh SYNOPSIS
.Nm
.Op Fl n Ar name
.Fl r Ar type
.Nm
.Op Fl n Ar name
.Fl R Ar index
.Nm
.Op Fl n Ar name
.Fl w Ar type
.Nm
.Op Fl n Ar name
.Fl l Op Fl v Op Fl v Op Fl v
.Nm
.Op Fl n Ar name
.Fl C
.Nm
.Op Fl n Ar name
.Fl c
.Nm
.Fl n Ar name
.Fl d
.Sh DESCRIPTION
The
.Nm
read and write content of the pasteboard.  When reading from the pasteboard,
content of specified type is written on stdout.  When writing
Pasteboard, data should be given via stdin.  It also displays information
about what the pasteboard contains.
.Pp
Following options are available:
.Bl -tag -width Ds
.It Fl n Ar name
Use the private pasteboard named as
.Ar name
instead of the general (standard) pasteboard.  If there is no pasteboards with 
given name, one will be created.
.It Fl r Ar type
Read data from the pasteboard and write them to stdout.  Available 
.Ar types
are text, rtf, rtfd, tiff, png, pdf, html, etc. (Macintosh PICT is no longer supported).
.It Fl R Ar index
Read n-th data from the pasteboard and write the data to stdout.  .Ar index is the
number shown by "pbdutil -lvvv"
.It Fl w Ar type
Read data from stdin and write them out to Pasteboard.  
.Ar type
and the data must match.
.It Fl l Op Fl v Op Fl v Op Fl v
List contents of Pasteboard briefly.  More 
.Fl v
make the result more specific.  "-lvvv" also shows index number that can
be used with "-R index" option.
.It Fl C
Count contents of the pasteboard
.It Fl c 
Clear contents of the pasteboard.
.It Fl n Ar name Fl d
Release the resources of the named pasteboard.  The name of the pasteboard must be specified as "-n name" and you cannot release the General (default) pasteboard.
.El
.Pp
.Sy mkfw is a companion program that makes rtfd structured directory from the output of pbdutil.  If the pasteboard contains data of RTFD type, you can use mkfw.
.\" .Sh IMPLEMENTATION NOTES
.Sh EXAMPLES
To read and write data in Pasteboard: 
.Pp
.Dl "pbdutil -r tiff > /tmp/a.tiff"
.Dl "pbdutil -w text < /etc/hosts"
.Pp
To clear data in the pasteboard:
.Pp
.Dl "pbdutil -c"
.Pp
To make rtfd structured directory that can be opened with TextEdit.app:
.Pp
.Dl "pbdutil -r rtfd | mkfw FILE.rtfd"
.Sh BUGS
There is no way to know private pasteboards' names.
.Pp
More than one type of data cannot be stored into Pasteboard.  For example,
when you copy a page from Safari, then Pasteboard contains both text and rtf
data.  But you cannot store both text and rtf data using pbdutil.  You can 
store only one type of data using pbdutil. 
.Pp
Now, private pasteboards created by issuing "pbdutil -n name" can be removed by using "-d" option with "-n name".  However, you have to remember the names of pasteboards that you created with "-n name" because there is still no way to know the names of existing private pasteboards.
.Sh SEE ALSO
.Pp
.Xr pbpaste 1 ,
.Xr pbcopy 1
.Sh AUTHOR
CHOI Kyong-Rok.


pbdutil.m
/* pbdutil.m */
/* Utility to read/write Pasteboard */
/* written by rok (CHOI Kyong-Rok) */
/* (C) by CHOI Kyong-Rok */

#import <Cocoa/Cocoa.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/uio.h>
#include <unistd.h>

void init();
void usage();
void listTypes(NSPasteboard *pbd, int verboseLevel);
BOOL writeDataForType(NSPasteboard *pbd, char *typename);
BOOL readDataForType(NSPasteboard *pbd, char *typename);
BOOL readNthData(NSPasteboard *pbd, int n);
void pbdclear(NSPasteboard *pbd);
void pbdcount(NSPasteboard *pbd, int verboseLevel);
void destroyPasteboard(NSPasteboard *pbd);

enum pbdmode {pbdutil, pbcopy, pbpaste, pbclear};

NSDictionary *pbTypes;

int
main(int argc, char *argv[])
{
    @autoreleasepool{
    int verboseLevel = 0;
    char *typename = NULL;
    NSPasteboard *pbd = nil;
    enum {get, set, list, help, clear, count, getNth, destroy, none} what = none;
    enum pbdmode pbdmode = pbdutil;
    char sw;
    int nth;

    init();	// setup pbTypes etc.

    NSString *cmd = [[NSString stringWithUTF8String: argv[0]] lastPathComponent];

    if([cmd compare: @"pbcopy"] == NSOrderedSame){
	pbdmode = pbcopy;
    }else if([cmd compare: @"pbpaste"] == NSOrderedSame){
	pbdmode = pbpaste;
    }else if([cmd compare: @"pbclear"] == NSOrderedSame){
        pbdmode = pbclear;
    }else{
	pbdmode = pbdutil;
    }

    if((pbdmode == pbcopy) || (pbdmode == pbpaste) || (pbdmode == pbclear)){
	// pbcopy / pbpaste compatible mode
	switch(pbdmode){
	    case pbcopy:
		what = set;
		break;
	    case pbpaste:
		what = get;
		break;
            case pbclear:
                what = clear;
                break;
	    case pbdutil:
	    default:
		// this must not happens
		break;
	}
	NSUserDefaults *args = [[NSUserDefaults alloc] init];
	if([args stringForKey: @"pboard"]){
	    pbd = [NSPasteboard pasteboardWithName:
		    [args stringForKey: @"pboard"]];
	}
	if([args stringForKey: @"Prefer"]){
	    typename = (char *)[[args stringForKey: @"Prefer"] UTF8String];
	}else{
	    typename = "text";
	}
    }else{
	while((sw = getopt(argc, argv, "vcChlw:r:R:n:o:d")) != -1){
	    switch(sw){
		case 'h':		// show help
		    what = help;
		    break;
		case 'l':		// list available types and information
		    what = list;
		    break;
		case 'w':		// write to pasteboard
		    what = set;
		    typename = optarg;
		    break;
		case 'r':		// read from pasteboard
		    what = get;
		    typename = optarg;
		    break;
		case 'c':		// clear pasteboard
		    what = clear;
		    break;
		case 'C':		// count the types of contents
		    what = count;
		    break;
		case 'v':
		    verboseLevel += 1;
		    break;
		case 'R':		// read n-th data from pasteboard
		    what = getNth;
		    nth = atoi(optarg);
		    break;
		case 'n':
		    pbd = [NSPasteboard pasteboardWithName:
				[NSString stringWithUTF8String: optarg]];
		    break;
                case 'd':
                    what = destroy;
                    break;
		default:
		    break;
	    }
	}
	argc -= optind;
	argv += optind;
    }

    if(pbd == nil)
	pbd  = [NSPasteboard generalPasteboard];

    switch(what){
	case help:
	    usage();
	    break;
	case get:
	    if(readDataForType(pbd, typename) != YES)
		usage();	// TYPE typename not exists
	    break;
	case set:
	    if(writeDataForType(pbd, typename) != YES)
		usage();	// TYPE typename not exists
	    break;
	case clear:
	    pbdclear(pbd);
	    break;
	case count:
	    pbdcount(pbd, verboseLevel);
	    break;
	case getNth:
	    readNthData(pbd, nth);
	    break;
        case destroy:
            destroyPasteboard(pbd);
            break;
	case list:		// Default Action
	default:
	    listTypes(pbd, verboseLevel);
	    break;
    }

    return 0;
    } // @autoreleasepool
}

void
init()
{
    pbTypes =
        @{
            @"text": NSPasteboardTypeString,
            @"txt": NSPasteboardTypeString,
            @"tiff": NSPasteboardTypeTIFF,
            @"png": NSPasteboardTypePNG,
            @"pdf": NSPasteboardTypePDF,
            @"html": NSPasteboardTypeHTML,
            @"rtf": NSPasteboardTypeRTF,
            @"rtfd": NSPasteboardTypeRTFD,
            @"tab": NSPasteboardTypeTabularText,
            @"url": NSPasteboardTypeURL,
            @"path": NSPasteboardTypeFileURL,
            @"font": NSPasteboardTypeFont
        };
}

void usage()
{
    printf("Usage:	pbdutil [-n name] [-r type|-R index|-w type|-l[ -v[ -v[ -v]]]|-c]\n");
    printf("	(supported types:");
    // output order may be strange
    [pbTypes enumerateKeysAndObjectsUsingBlock:
        ^(NSString *k, NSString *obj, BOOL *stop){
            printf(" %s", [k UTF8String]);
        }
    ];
    printf(")\n");
    printf("	pbdutil -C\n");
    printf("	pbdutil -n name -d\n");
    exit(1);
}

void
listTypes(NSPasteboard *pbd, int verboseLevel)
{
    NSString *t;
    NSArray *types = [pbd types];
    int i = 1;

    if([types count] == 0) return;
    printf("Available type(s):");
    for(t in types){
	if(verboseLevel <= 2){
            [pbTypes enumerateKeysAndObjectsUsingBlock:^(NSString *key, NSString *value, BOOL *stop){
                if([t isEqualToString: value]){
		    if(verboseLevel <= 0){
			printf("\t%s", (char *)[key UTF8String]);
		    }else if(verboseLevel >= 1){
			printf("\n\t%s (%s)", (char *)[key UTF8String],
				[t UTF8String]);
			if(verboseLevel >= 2){
			    printf(" (size: %lu bytes)",
				(unsigned long)[[pbd dataForType: t] length]);
			}
		    }
                }
            }];
	}else{	// verboseLevel >= 3
		// list all types including non-supported types
	    printf("\n\t%2d: %s (size: %lu)", i++, [t UTF8String], (unsigned long)[[pbd dataForType: t] length]);
	}
    }
    printf("\n");
}

// write data in Pasteboard to stdout
BOOL
readDataForType(NSPasteboard *pbd, char *typename)
{
    NSString *type;

    if((type = [pbTypes objectForKey:
	    [NSString stringWithUTF8String: typename]]) != nil){
	NSData *t = [pbd dataForType: type];
	if(t != nil){	// t should not be nil
	    int l = [t length];
	    void *d = (void *)[t bytes];
	    int c = 0;
	    while(l > 0){
		c = write(1, d, l>256?256:l);
		if(c>=0){
		    d += c;
		    l -= c;
		}else{
		    // write(2) failed
		    exit(-1);
		}
	    }
	}
    }else{
	return NO;
    }
    return YES;
}

// write n-th data in Pasteboard to stdout
BOOL
readNthData(NSPasteboard *pbd, int n)
{
    NSArray *types = [pbd types];

    if((n < 1) || ([types count] < n)){
	return FALSE;
    }

    NSData *t = [pbd dataForType: [types objectAtIndex: n -1]];
    if(t != nil){	// t should not be nil
	int l = [t length];
	void *d = (void *)[t bytes];
	int c = 0;
	while(l > 0){
	    c = write(1, d, l>256?256:l);
	    if(c>=0){
		d += c;
		l -= c;
	    }else{
		// write(2) failed
		exit(-1);
	    }
	}
    }
    
    return YES;
}


// read data from stdin and write to Pasteboard
BOOL
writeDataForType(NSPasteboard *pbd, char *typename)
{
    int l;
    char buffer[256];
    NSString *type;
    NSMutableData *data = [NSMutableData dataWithLength: 0];
#if DEBUG > 1
    int readCount = 0;
#endif

    if((type = [pbTypes objectForKey:
	    [NSString stringWithUTF8String: typename]]) != nil){
	while((l = read(0, buffer, 256)) > 0){
	    [data appendBytes: buffer length: l];
#if DEBUG > 1
	    readCount += l;
#endif
	}
	if(l < 0){		// read error
	    perror("Read error: ");
	    return NO;
	}
#if DEBUG > 1
	fprintf(stderr, "%d bytes read.\n", 
#endif
	[pbd declareTypes: [NSArray arrayWithObject: type] owner: nil];
	[pbd setData: data forType: type];
    }else{
    	return NO;
    }

    return YES;
}

void
pbdclear(NSPasteboard *pbd)
{
    NSString *nm = [pbd name];
    if([nm isEqualToString: NSPasteboardNameGeneral]
	|| [nm isEqualToString: NSPasteboardNameFont]
	|| [nm isEqualToString: NSPasteboardNameRuler]
	|| [nm isEqualToString: NSPasteboardNameFind]
	|| [nm isEqualToString: NSPasteboardNameDrag]){
	(void)[pbd clearContents];
    }else{
#if DEBUG > 1
	NSLog(@"Clear Private Pasteboard: %s\n", [nm UTF8String]);
#endif
	(void)[pbd clearContents];
	(void)[pbd releaseGlobally];
    }
    return;
}

void
pbdcount(NSPasteboard *pbd, int verboseLevel)
{
    printf("%lu", (unsigned long)[[pbd types] count]);
    if(verboseLevel > 0)
       printf(" types are available");	
    printf("\n");
}

void
destroyPasteboard(NSPasteboard *pbd)
{
    if(pbd == [NSPasteboard generalPasteboard]){
        fprintf(stderr, "You cannot destroy general (default) Pasteboard\n");
        exit(1);
    }
    if(pbd){
        [pbd releaseGlobally];
    }
    return;
}
